### PR TITLE
Add dependency functionality to config (#31)

### PR DESCRIPTION
**Pull request by Kotarski**
Lots of configuration infra to link paths output from transforms to paths input into other transforms and config values

This is not used yet but is a stepping stone to getting workflows working properly.

The next thing will be to extend the `config.Linker` to calculate dependencies between transforms (most of the way there) and then we can run the transforms in order (until later when we need to deal with parallelism and it gets a bit more complicated)

### CODE REVIEW

File: blockwork/config/base.py [hunk 1]
```diff
@@ -12,20 +12,88 @@
...
 class Project:
     'Base class for project configuration'
     blocks: dict[str, str]
 
-@dataclass
+
+@dataclass(kw_only=True)
+class ElementContext:
+    'Context object bound on to each element to keep track of where it came from'
+    block: str
```
Comment (Intuity) [inline blockwork/config/base.py:41]: Unit instead of block?

File: blockwork/config/config.py [hunk 2]
```diff
@@ -1,6 +1,90 @@
+from pathlib import Path
+from typing import Iterable, cast, overload, TYPE_CHECKING
 from blockwork.common.registry import RegisteredClass
```
Comment (Intuity) [inline blockwork/config/config.py:3]: Why is this using absolute imports?

File: blockwork/config/config.py [hunk 3]
```diff
@@ -1,6 +1,90 @@
...
+    @ed.kotarski: The next extension to this will be to use that
+    information to create dependencies between transforms to determine 
+    the order they need to be run.
+    '''
+    def __init__(self, config: Config):
+        self.config = config
+        self.pairs: dict[tuple[str, str], Path] = {}
+
+    def link_inputs(self, element: base.Element):
+        block = cast(base.ElementContext, element._context).block
```
Comment (Intuity) [inline blockwork/config/config.py:47]: Guess this becomes unit as well

File: blockwork/config/config.py [hunk 4]
```diff
@@ -1,6 +1,90 @@
...
+        def resolve(paths: list[str]) -> list[str]: ...
+        def resolve(paths):
+            if (is_single := isinstance(paths, str)):
+                paths = [paths]
+
+            resolved_paths: list[str] = []
+            for path in paths:
+                full_path = block_scratch / path
+                self.pairs[(block, path)] = full_path
+                resolved_paths.append(full_path.as_posix())
```
Comment (Intuity) [inline blockwork/config/config.py:86]: Again could yield?

File: example/bench/testbench.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+!Testbench
+design: !Design top
+bench_python: bench/bench.py
```
Comment (Intuity) [inline example/bench/testbench.yaml:3]: Could we eventually tier these into:

```
!Testbench
design: !Design [top]
bench:
  python: ...
  make: ...
```

**Review by Intuity — APPROVED** [review]

File: blockwork/config/config.py (hunk 4 above)
Comment (Kotarski) [reply blockwork/config/config.py:86]: Discussed - would make more complex for caller

File: blockwork/config/config.py (hunk 3 above)
Comment (Kotarski) [reply blockwork/config/config.py:47]: Agreed

File: example/bench/testbench.yaml (hunk 5 above)
Comment (Kotarski) [reply example/bench/testbench.yaml:3]: Yep!

File: blockwork/config/config.py (hunk 2 above)
Comment (Kotarski) [reply blockwork/config/config.py]: Fixed this instance, will fix others in follow-up

**Pull request merged**